pkg/reconciler/reconciler.go

```diff
@@ -0,0 +1,75 @@
+package reconciler
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type Reconciler interface {
+	Reconcile(ctx context.Context, c client.Client, scheme *runtime.Scheme) error
+}
+
+type UpdateReconciler struct {
+	controller metav1.Object
+	resource   client.Object
+}
+
+func (r UpdateReconciler) Reconcile(ctx context.Context, c client.Client, scheme *runtime.Scheme) error {
```

Comment:
I think it would be nice for all `Reconcile` to return `ctrl.Result, error` so that each reconcile method can send back to the caller that the request just needs to be requeued.

Reply:
Right now the reconciler doesn't know anything about how `controller-runtime` (re)schedules events. To me that decision belongs in the controller.
Logging happens there, as well as deciding how long to hold up the requeued event. That is nothing the reconciler needs to have knowledge about.

Reply:
I kind of agree with on the requirement to remove coupling however, I do feel that the reconcile method should be able to not just `Patch` or `Delete` but also be able to communicate back to the "controller / caller" that the request need to be retried without an error ( i.e. the controller should return `Result{Requeue : true, After 5 * time.Second }`.  While in other cases return an error just how it does now. 

I am open to changing this in future when such a requirement arise.